File: MaxCut.py
import random as rand
import load_input

vertices, Edges = load_input.max_cut("input_format")


def checkEdge(S, T):
    Sum = 0
    for (ai, aj), w in Edges:
        if (ai in S and aj in T) or (ai in T and aj in S):
            Sum += w
    return Sum


def checkNeighbors(S, T, solution):
    Sn = S.copy()
    Tn = T.copy()
    Schosen = Sn.copy()
    Tchosen = Tn.copy()
    for a in S:
        Sn.remove(a)
        Tn.append(a)
        new_solution = checkEdge(Sn, Tn)
        # print(f"Neighbor partition: {Sn} | {Tn}")
        # print(f"Current cut value: {new_solution}")
        if solution <= new_solution:
            Schosen = Sn.copy()
            Tchosen = Tn.copy()
            solution = new_solution
        Sn = S.copy()
        Tn = T.copy()
    for a in T:
        Tn.remove(a)
        Sn.append(a)
        new_solution = checkEdge(Sn, Tn)
        # print(f"Neighbor partition: {Sn} | {Tn}")
        # print(f"Current cut value: {new_solution}")
        if solution <= new_solution:
            Schosen = Sn.copy()
            Tchosen = Tn.copy()
            solution = new_solution
        Sn = S.copy()
        Tn = T.copy()
    return Schosen, Tchosen, solution


def random_iteration(num_vertices):
    rand.shuffle(vertices)
    S = []
    T = []
    # Split at the second place
    for i in range(num_vertices):
        if i < 3:
            S.append(vertices[i])
        else:
            T.append(vertices[i])
    # Check for Cuts on edges
    Schosen = S.copy()
    Tchosen = T.copy()
    # print(f"Current partition: {S} | {T}")
    # print(f"Current cut value: {checkEdge(S, T)}")
    solution = checkEdge(S, T)
    # Check for neighbors:
    improved = True
    highestSolution = solution
    while improved:
        # print("========== Climbing ?")
        improved = False
        Schosen, Tchosen, highestSolution = checkNeighbors(
            S, T, solution)
        if highestSolution > solution:
            improved = True
            # print(f"-- Yes, Climbing")
            S = Schosen.copy()
            T = Tchosen.copy()
            solution = highestSolution
        else:
            # print(f"-- No, :(")
            break
    return highestSolution


HighestSolution = 0

for i in range(150):
    solution = random_iteration(len(vertices))
    if HighestSolution < solution:
        HighestSolution = solution
print(f"{HighestSolution}")

File: MaxCut.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

struct Edge {
    string u, v;
    int weight;
};

struct Graph {
    vector<string> vertices;
    vector<Edge> edges;
};

// Function to parse the input file
Graph loadInput(const string& filename) {
    Graph g;
    ifstream file(filename);
    string line, mode;

    while (getline(file, line)) {
        if (line.empty()) continue;
        if (line == "V" || line == "E") {
            mode = line;
            continue;
        }

        if (mode == "V") {
            // Remove parentheses
            line.erase(remove(line.begin(), line.end(), '('), line.end());
            line.erase(remove(line.begin(), line.end(), ')'), line.end());
            g.vertices.push_back(line);
        } else if (mode == "E") {
            stringstream ss(line);
            string u, v, w_str;
            getline(ss, u, '-');
            getline(ss, v, '-');
            getline(ss, w_str);

            // Clean up node names
            u.erase(remove(u.begin(), u.end(), '('), u.end());
            u.erase(remove(u.begin(), u.end(), ')'), u.end());
            v.erase(remove(v.begin(), v.end(), '('), v.end());
            v.erase(remove(v.begin(), v.end(), ')'), v.end());

            g.edges.push_back({u, v, stoi(w_str)});
        }
    }
    return g;
}

// Calculate the weight of the cut
long long checkEdge(const unordered_set<string>& S,
                    const unordered_set<string>& T, const vector<Edge>& edges) {
    long long sum = 0;
    for (const auto& edge : edges) {
        if ((S.count(edge.u) && T.count(edge.v)) ||
            (T.count(edge.u) && S.count(edge.v))) {
            sum += edge.weight;
        }
    }
    return sum;
}

// Local search: Try moving one node to the other set to improve the cut
long long checkNeighbors(unordered_set<string>& S, unordered_set<string>& T,
                         long long currentSolution, const vector<Edge>& edges) {
    long long bestLocalSolution = currentSolution;
    string bestNode = "";
    bool fromS = true;

    // Try moving nodes from S to T
    for (const string& node : S) {
        unordered_set<string> nextS = S;
        unordered_set<string> nextT = T;
        nextS.erase(node);
        nextT.insert(node);

        long long newSol = checkEdge(nextS, nextT, edges);
        if (newSol > bestLocalSolution) {
            bestLocalSolution = newSol;
            bestNode = node;
            fromS = true;
        }
    }

    // Try moving nodes from T to S
    for (const string& node : T) {
        unordered_set<string> nextS = S;
        unordered_set<string> nextT = T;
        nextT.erase(node);
        nextS.insert(node);

        long long newSol = checkEdge(nextS, nextT, edges);
        if (newSol > bestLocalSolution) {
            bestLocalSolution = newSol;
            bestNode = node;
            fromS = false;
        }
    }

    // Apply the best move found
    if (!bestNode.empty()) {
        if (fromS) {
            S.erase(bestNode);
            T.insert(bestNode);
        } else {
            T.erase(bestNode);
            S.insert(bestNode);
        }
    }

    return bestLocalSolution;
}

long long randomIteration(Graph& g, mt19937& rng) {
    vector<string> shuffledV = g.vertices;
    shuffle(shuffledV.begin(), shuffledV.end(), rng);

    unordered_set<string> S, T;
    for (int i = 0; i < shuffledV.size(); ++i) {
        if (i < 3)
            S.insert(shuffledV[i]);
        else
            T.insert(shuffledV[i]);
    }

    long long currentSolution = checkEdge(S, T, g.edges);
    bool improved = true;

    while (improved) {
        long long nextSolution = checkNeighbors(S, T, currentSolution, g.edges);
        if (nextSolution > currentSolution) {
            currentSolution = nextSolution;
        } else {
            improved = false;
        }
    }
    return currentSolution;
}

int main() {
    Graph g = loadInput("input_format");  // Ensure your filename matches
    if (g.vertices.empty()) {
        cout << "No vertices found. Check input file path/format." << endl;
        return 1;
    }

    random_device rd;
    mt19937 rng(rd());

    long long highestSolution = 0;

    cout << "Running 100 iterations..." << endl;
    for (int i = 0; i < 100; ++i) {
        long long sol = randomIteration(g, rng);
        if (sol > highestSolution) {
            highestSolution = sol;
        }
    }

    cout << "Highest Max-Cut found: " << highestSolution << endl;

    return 0;
}